keps/sig-cluster-lifecycle/kubeadm/20190424-kubeadm-for-windows.md

```diff
@@ -0,0 +1,293 @@
+
+After these steps, the node status should be Ready.
+
+### Implementation Details/Notes/Constraints
+
+#### Kubeadm manages the kubelet start / stop as a service
+
+Kubeadm has a “kubelet” phase that can start/stop the [kubelet service](https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/kubelet/kubelet.go).
+
+Kubeadm already supports running the kubelet as a Windows service. This support was added by these two PRs: [60144](https://github.com/kubernetes/kubernetes/pull/60144), [60144](https://github.com/kubernetes/kubernetes/pull/53553).
+
+This existing support will be leveraged by this proposal.
+
+#### Kubeadm makes assumptions about systemd and Linux
+
+Kubeadm is quite coupled to systemd and Linux at the moment.
+This file shows some of the kubelet-flags that are Linux bound:
+[kubelet flags](https://github.com/kubernetes/kubernetes/blob/b8b689aae0d6879de5192b203590330a11c7b9e3/cmd/kubeadm/app/phases/kubelet/flags.go)
+
+The following changes are required::
+* Implement proper cri-socket detection in kubeadm for Windows.
+* Omit flags such as “cgroup-driver”
+* Specify the correct “pod-infra-container-image”
+* Also omit writing systemd drop-in files on Windows in general.
+* Populate the Windows service dynamically.
+
+To support Windows specific flags for the kubelet, there is a requirement to split kubeadm’s app/phases/kubelet/flags.go files into two:
+* app/phases/kubelet/flags_windows.go
+* app/phases/kubelet/flags_linux.go
+
+In the case of Windows, we will be updating the registration of the kubelet as a Windows service, by changing its flags based on the flags kubeadm wants to use, complying with Windows specific defaults in kubeadm.
+
+The existing process of passing the following files to the kubelet, will be leveraged:
+* /etc/kubernetes/bootstrap-kubelet.conf
+* /etc/kubernetes/kubelet.conf
+* /var/lib/kubelet/config.yaml
+
+Windows related adjustments to default paths might be required.
+
+#### Windows vs Linux host paths
+
+Kubeadm makes a number of non-portable assumptions about paths. E.g. “/etc/kubernetes” is a hardcoded path in kubeadm.
+
+We need to evaluate the kubeadm codebase for such instances of non-portable paths - CRI sockets, Cert paths, etc. Such paths need to be defaulted properly in the kubeadm configuration API.
+
+A consolidated list of paths where kubeadm installs files to a set of paths needs to be created and updated to comply with the Windows OS model. At least a single PR against kubeadm will be required to modify the Windows defaults. 
+
+Last, as new paths are created, restrictive Access Control Lists (ACL) for Windows should be applied. Golang does not convert Posix permissions to an appropriate Windows ACL, so an additional step is needed. See [mkdirWithACL from moby/moby](https://github.com/moby/moby/blob/e4cc3adf81cc0810a416e2b8ce8eb4971e17a3a3/pkg/system/filesys_windows.go#L103)) for an example. This step will be performed by the wrapper script.
```

Comment:
This can be done by kubeadm itself, I don't think there's a need to do it from the wrapper

Reply:
this part is conflating
cc @ksubrmnn 

we want to minimize the changes in kubeadm, because the maintainers have expressed that the lack of kube-proxy and CNI as DaemonSets is quite "alpha".
if it's simply not feasible to do it outside in the wrapper, a change in kubeadm will be considered.

Reply:
My point is that even if we got privileged containers on Windows these concerns about file permissions would remain valid, so we should be putting them into kubeadm. The wrapper should only be for the things that are working around the lack of privileged support.

Reply:
I still share @benmoss 's opinion that this needs to be done by kubeadm itself.